pymc3/step_methods/ATMCMC.py

```diff
@@ -0,0 +1,499 @@
+                    q = (q0 + delta).astype(int)
+                else:
+                    delta[self.discrete] = round(
+                                    delta[self.discrete], 0).astype(int)
+                    q = q0 + delta
+                    q = q[self.discrete].astype(int)
+            else:
+                q = q0 + delta
+
+            if self.check_bnd:
+                varlogp = self.check_bnd(q)
+                if num.isfinite(varlogp):
+                    q_new = pm.metropolis.metrop_select(
+                                    self.beta * self.delta_logp(q, q0), q, q0)
+                    if q_new is q:
+                        self.accepted += 1
+                else:
+                    q_new = q0
+            else:
+                q_new = pm.metropolis.metrop_select(
+                                self.beta * self.delta_logp(q, q0), q, q0)
+                if q_new is q:
+                    self.accepted += 1
+
+            self.steps_until_tune -= 1
+        return q_new
+
+    def calc_beta(self):
+        '''Calculate next tempering beta and importance weights based on
+           current beta and sample likelihoods.
+            returns: beta(m+1),
+                     beta(m),
+                     NdArray of importance weights (floats)'''
+
+        low_beta = self.beta
+        up_beta = 2.
+        old_beta = self.beta
+
+        while up_beta - low_beta > 1e-6:
+            current_beta = (low_beta + up_beta) / 2.
+            temp = num.exp((current_beta - self.beta) * \
+                           (self.likelihoods - self.likelihoods.max()))
+            cov_temp = num.std(temp) / num.mean(temp)
+            if cov_temp > 1:
+                up_beta = current_beta
+            else:
+                low_beta = current_beta
+
+        beta = current_beta
+        weights = temp / num.sum(temp)
+        return beta, old_beta, weights
+
+    def calc_covariance(self):
+        '''Calculate trace covariance matrix based on importance weights.'''
+        return num.cov(self.array_population, aweights=self.weights, rowvar=0)
+
+    def select_end_points(self, mtrace):
+        '''Read trace results and take end points for each chain and set as
+           start population for the next stage.
+            Input: Multitrace object
+            Returns: population,
+                     array_population,
+                     likelihoods'''
+        array_population = num.zeros((self.N_chains,
+                                      self.ordering.dimensions))
+        N_steps = len(mtrace)
+        bij = pm.DictToArrayBijection(self.ordering, self.population[0])
+
+        if self.stage > 0:
+            # collect end points of each chain and put into array
+            for var, slc, shp, _ in self.ordering.vmap:
+                if len(shp) == 0:
+                    array_population[:, slc] = num.atleast_2d(
+                        mtrace.get_values(
+                                    varname=var,
```

Comment:
Maybe you are confused by my line breaks? If above linelength of 80 I start a new line ... because of the width of a standard A4 page.